includes/payment-methods/class-wc-stripe-express-checkout-element.php

```
@@ -110,6 +113,147 @@ public function init() {
+			return;
+		}
+
+		$errors = new WP_Error();
+		/**
+		 * Allow third-party plugins to validate custom checkout data for express checkout orders.
+		 *
+		 * To be used as a stand-in for the `woocommerce_after_checkout_validation` action.
+		 *
+		 * @since 9.6.0
+		 *
+		 * @param array $custom_checkout_data The custom checkout data.
+		 * @param WP_Error $errors The WP_Error object, for adding errors when validation fails.
+		 */
+		do_action( 'wc_gateway_stripe_express_checkout_after_checkout_validation', $custom_checkout_data, $errors );
+
+		if ( $errors->has_errors() ) {
+			$error_messages = implode( "\n", $errors->get_error_messages() );
+			throw new WC_Data_Exception( 'wc_gateway_stripe_express_checkout_invalid_data', $error_messages );
+		}
+
+		/**
+		 * Allow third-party plugins to add custom checkout data for express checkout orders.
+		 *
+		 * To be used as a stand-in for the `woocommerce_checkout_update_order_meta` action.
+		 *
+		 * @since 9.6.0
+		 *
+		 * @param integer $order_id The order ID.
+		 * @param array $custom_checkout_data The custom checkout data.
+		 */
+		do_action( 'wc_gateway_stripe_express_checkout_update_order_meta', $order->get_id(), $custom_checkout_data );
+	}
+
+	/**
+	 * Get custom checkout data from the request object.
+	 *
+	 * To support custom fields in express checkout and classic checkout,
+	 * we pass custom data as extended data, i.e. under extensions.
+	 *
+	 * @param WP_REST_Request $request The request object.
+	 * @return array Custom checkout data.
+	 */
+	private function get_custom_checkout_data_from_request( $request ) {
+		$extensions = $request->get_param( 'extensions' );
+		if ( empty( $extensions ) ) {
+			return [];
+		}
+
+		$custom_checkout_data_json = $extensions['wc-stripe/express-checkout']['custom_checkout_data'] ?? '';
+		if ( empty( $custom_checkout_data_json ) ) {
+			return [];
+		}
+
+		$custom_checkout_data = json_decode( $custom_checkout_data_json, true );
```

Perhaps we should perform some basic validation on the decoded data: verify that it is a valid array or check for decoding errors. Should we return the default value of throw an exception when it is invalid?

---

Added validation to check that the result is not empty and an array.